### PR TITLE
CM-1012: How to Determine If a Key is Pressed in a .NET Console Application

### PR DESCRIPTION
**Pull request by mohitkumar342**

### CODE REVIEW

**Comment by mohitkumar342** [comment]
Title updated as per feedback

**Review by elfico — CHANGES_REQUESTED** [review]

**Review by mohitkumar342 — COMMENTED** [review]
Changes implemented

**Comment by mohitkumar342** [comment]
Changes implemented

File: csharp-basic-topics/DetectKeyPressInCSharp/Tests/Tests.csproj [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    <IsTestProject>true</IsTestProject>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.6.0" />
+    <PackageReference Include="Moq" Version="4.20.70" />
+    <PackageReference Include="MSTest.TestAdapter" Version="3.0.4" />
+    <PackageReference Include="MSTest.TestFramework" Version="3.0.4" />
+    <PackageReference Include="coverlet.collector" Version="6.0.0" />
+    <PackageReference Include="NUnit" Version="4.0.1" />
```
Comment (elfico) [inline csharp-basic-topics/DetectKeyPressInCSharp/Tests/Tests.csproj:18]: The tests are not running in VS. 
Installing `NUnit3TestAdapter` nuget should make it work

**Review by elfico — CHANGES_REQUESTED** [review]

File: csharp-basic-topics/DetectKeyPressInCSharp/DetectKeyPressInCSharp/DetectKeyPressInCSharp/ConsoleApp.cs [hunk 2]
```diff
@@ -0,0 +1,40 @@
+﻿using System;
```
[the remark this answers is not in the document]
Comment (mohitkumar342) [reply csharp-basic-topics/DetectKeyPressInCSharp/DetectKeyPressInCSharp/DetectKeyPressInCSharp/ConsoleApp.cs]: Removed

File: csharp-basic-topics/DetectKeyPressInCSharp/Tests/ConsoleAppTests.cs [hunk 3]
```diff
@@ -0,0 +1,52 @@
+using DetectKeyPressInCSharp;
+using Moq;
+using NUnit.Framework;
+using System.Runtime;
```
[the remark this answers is not in the document]
Comment (mohitkumar342) [reply csharp-basic-topics/DetectKeyPressInCSharp/Tests/ConsoleAppTests.cs]: Done

**Comment by mohitkumar342** [comment]
Changes done

**Review by elfico — APPROVED** [review]

File: csharp-basic-topics/DetectKeyPressInCSharp/DetectKeyPressInCSharp/DetectKeyPressInCSharp/ConsoleApp.cs [hunk 4]
```diff
@@ -0,0 +1,34 @@
+﻿namespace DetectKeyPressInCSharp
+{
+    public class ConsoleApp
+    {
+        private readonly IConsoleService _consoleService;
+        public ConsoleApp(IConsoleService consoleService)
```
Comment (elfico) [inline csharp-basic-topics/DetectKeyPressInCSharp/DetectKeyPressInCSharp/DetectKeyPressInCSharp/ConsoleApp.cs:6]: @mohitkumar342 
There should be a space here

File: csharp-basic-topics/DetectKeyPressInCSharp/Tests/ConsoleAppTests.cs [hunk 5]
```diff
@@ -0,0 +1,50 @@
+using DetectKeyPressInCSharp;
+using Moq;
+using NUnit.Framework;
+
+namespace Tests
+{ 
+    [TestFixture]
+    public class ConsoleAppTests
+    {
+        private Mock<IConsoleService> mockConsoleService;
```
Comment (elfico) [inline csharp-basic-topics/DetectKeyPressInCSharp/Tests/ConsoleAppTests.cs:10]: @mohitkumar342 
Both private fields  should be prefixed with _

File: csharp-basic-topics/DetectKeyPressInCSharp/Tests/ConsoleAppTests.cs [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+        private ConsoleApp app;
+
+        [SetUp]
+        public void Setup()
+        {
+            mockConsoleService = new Mock<IConsoleService>();
+            app = new ConsoleApp(mockConsoleService.Object);
+        }
+
+        [Test]
```
Comment (elfico) [inline csharp-basic-topics/DetectKeyPressInCSharp/Tests/ConsoleAppTests.cs:20]: Also, test names should be following Given_When_Then.

Comment (elfico) [inline csharp-basic-topics/DetectKeyPressInCSharp/Tests/ConsoleAppTests.cs:20]: Also, test names should be following Given_When_Then. @mohitkumar342

Comment (elfico) [inline csharp-basic-topics/DetectKeyPressInCSharp/Tests/ConsoleAppTests.cs:20]: Also, test names should be following Given_When_Then. @mohitkumar342

**Review by elfico — CHANGES_REQUESTED** [review]

File: csharp-basic-topics/DetectKeyPressInCSharp/DetectKeyPressInCSharp/DetectKeyPressInCSharp/ConsoleApp.cs (hunk 4 above)
Comment (mohitkumar342) [reply csharp-basic-topics/DetectKeyPressInCSharp/DetectKeyPressInCSharp/DetectKeyPressInCSharp/ConsoleApp.cs:7]: Done

File: csharp-basic-topics/DetectKeyPressInCSharp/Tests/ConsoleAppTests.cs (hunk 6 above)
Comment (mohitkumar342) [reply csharp-basic-topics/DetectKeyPressInCSharp/Tests/ConsoleAppTests.cs:20]: DOne

Comment (mohitkumar342) [reply csharp-basic-topics/DetectKeyPressInCSharp/Tests/ConsoleAppTests.cs:20]: DOne

**Comment by mohitkumar342** [comment]
Changes are done, thanks for the valuable reviews.

File: csharp-basic-topics/DetectKeyPressInCSharp/Tests/ConsoleAppTests.cs [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+            _app.ReadKeyUse();
+
+            _mockConsoleService.Verify(m => m.WriteLine("Press any key to continue..."), Times.Exactly(2));
+            _mockConsoleService.Verify(m => m.WriteLine("\nYou pressed: A"), Times.Once());
+            _mockConsoleService.Verify(m => m.WriteLine("\nYou pressed: B"), Times.Once());
+            _mockConsoleService.Verify(m => m.WriteLine("Process stopped"), Times.Once());
+        }
+
+        [Test]
+        public void KeyInputs_KeyAvailableUseIsCalled_ShouldStopWhenXIsPressed()
```
Comment (elfico) [inline csharp-basic-topics/DetectKeyPressInCSharp/Tests/ConsoleAppTests.cs:36]: @mohitkumar342 the test name doesn't follow the format: Given_When_Then.
A better name would be `ApplicationIsRunning_WhenXIsPressed_ApplicationShouldStop` or something similar.

Comment (mohitkumar342) [reply csharp-basic-topics/DetectKeyPressInCSharp/Tests/ConsoleAppTests.cs]: Thank you, I have changed the Test Name as you suggested.

**Review by elfico — APPROVED** [review]

File: csharp-basic-topics/DetectKeyPressInCSharp/Tests/ConsoleAppTests.cs [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+        [SetUp]
+        public void Setup()
+        {
+            _mockConsoleService = new Mock<IConsoleService>();
+            _app = new ConsoleApp(_mockConsoleService.Object);
+        }
+
+        [Test]
+        public void KeyInputs_ReadKeyUseIsCalled_ShouldPrintCorrectMessages()
```
Comment (elfico) [inline csharp-basic-topics/DetectKeyPressInCSharp/Tests/ConsoleAppTests.cs:21]: @mohitkumar342 
Feedback from the senior editor is that `Our tests still aren’t following Given_When_Then.`
We can modify this as `WhenConsoleReadkey_ThenPrintCorrectMessage`

File: csharp-basic-topics/DetectKeyPressInCSharp/Tests/ConsoleAppTests.cs [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+            _app.ReadKeyUse();
+
+            _mockConsoleService.Verify(m => m.WriteLine("Press any key to continue..."), Times.Exactly(2));
+            _mockConsoleService.Verify(m => m.WriteLine("\nYou pressed: A"), Times.Once());
+            _mockConsoleService.Verify(m => m.WriteLine("\nYou pressed: B"), Times.Once());
+            _mockConsoleService.Verify(m => m.WriteLine("Process stopped"), Times.Once());
+        }
+
+        [Test]
+        public void ApplicationIsRunning_WhenXIsPressed_ApplicationShouldStop()
```
Comment (elfico) [inline csharp-basic-topics/DetectKeyPressInCSharp/Tests/ConsoleAppTests.cs:36]: @mohitkumar342 ,
we also need to make some adjustment here.
`WhenXKeyIsPressed_ThenApplicationShouldStop`

**Review by elfico — CHANGES_REQUESTED** [review]

File: csharp-basic-topics/DetectKeyPressInCSharp/Tests/ConsoleAppTests.cs (hunk 8 above)
Comment (mohitkumar342) [reply csharp-basic-topics/DetectKeyPressInCSharp/Tests/ConsoleAppTests.cs]: This is done,

File: csharp-basic-topics/DetectKeyPressInCSharp/Tests/ConsoleAppTests.cs (hunk 9 above)
Comment (mohitkumar342) [reply csharp-basic-topics/DetectKeyPressInCSharp/Tests/ConsoleAppTests.cs]: The test name is updated. Thanks

**Pull request merged**